util/src/MultiSizeBufferPool.cpp

```diff
@@ -0,0 +1,688 @@
+  TimeRecorder scoped_timer(*histograms_.time_in_reportBufferUsage);
+  if (usageSize == 0) {
+    throwException<std::invalid_argument>(logger_, "Invalid resultSize (must be non-zero)!", KVLOG(KVLOG_PREFIX));
+  }
+  if (auto [result, errMsg] = validateBuffer(buffer); !result) {
+    throwException<std::invalid_argument>(logger_, errMsg, KVLOG(KVLOG_PREFIX, (void*)buffer));
+  }
+
+  auto bufferSize = getBufferSize(buffer);
+  auto chunk = bufferToChunk(buffer);
+  getSubPool(bufferSize).reportBufferUsage(chunk, usageSize);
+  subPoolSelector_->reportBufferUsage(bufferSize, usageSize);
+  if (bufferSize >= usageSize) {
+    histograms_.caller_allocated_buffer_size->record(bufferSize);
+    histograms_.caller_allocated_buffer_usage->record(usageSize);
+  }
+}
+
+std::pair<char*, uint32_t> MultiSizeBufferPool::getBufferByMinBufferSize(uint32_t minBufferSize) {
+  TimeRecorder scoped_timer(*histograms_.time_in_getBufferByMinBufferSize);
+
+  // lower_bound returns an iterator to the 1st size in bufferSizes_ which is no less than  requested chunkSize
+  // if not such, it returns last iterator
+  DEBUG_PRINT(logger_, "getBufferByMinBufferSize enter:" << KVLOG(KVLOG_PREFIX));
+  auto lowestSizeToFitIter = bufferSizes_.lower_bound(minBufferSize);
+  if (lowestSizeToFitIter == bufferSizes_.end()) {
+    auto maxSubPoolBufferSize = *bufferSizes_.rbegin();
+    throwException<std::invalid_argument>(
+        logger_, "Invalid requested bufferSize:", KVLOG(KVLOG_PREFIX, minBufferSize, maxSubPoolBufferSize));
+  }
+  auto result = popBuffer(*lowestSizeToFitIter);
+  DEBUG_PRINT(logger_, "getBufferByMinBufferSize exit:" << KVLOG(KVLOG_PREFIX) << stats_);
+  return result;
+}
+
+void MultiSizeBufferPool::doPeriodic() {
+  // This is a timer callback: here the pool reports statistics and evaluate its state (for enabled attributes)
+  std::future<void> purgeFuture, statusReportFuture, histogramsReportFuture, metricsReportFuture;
+  ++periodicCounter_;
+  if ((config_.purgeEvaluationFrequency > 0) && (periodicCounter_ % config_.purgeEvaluationFrequency) == 0) {
+    purgeFuture = std::async(std::launch::async, [this]() {
+      purger_->evaluate(*this);  // TODO , an empty call, nothing passed yet and no API to pool defined.
+    });
+  }
+  if ((config_.statusReportFrequency > 0) && (periodicCounter_ % config_.statusReportFrequency) == 0) {
+    // Report statistics to log
+    statusReportFuture = std::async(std::launch::async, [this]() { this->statusReport(); });
+  }
+  if ((config_.histogramsReportFrequency > 0) && (periodicCounter_ % config_.histogramsReportFrequency) == 0) {
+    // report histograms to log
+    histogramsReportFuture = std::async(std::launch::async, [this]() {
+      auto& registrar = concord::diagnostics::RegistrarSingleton::getInstance();
+      registrar.perf.snapshot(name_);
+      LOG_INFO(logger_, "Histograms report:" << registrar.perf.toString(registrar.perf.get(name_)));
```

Comment:
in merge to master this log can be in debug level

Reply:
we currently print it once a minute, pretty rare.
If we don't want it at all we can just set histogramsReportFrequency=0
I guess it's a matter of taste...
but usually when you need the info many time you loose it since you put it in debug level...
once a minute is not so much.